File: wear/src/main/java/com/example/next_contest/wear/data/AuthRepository.java
package com.example.next_contest.wear.data;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

public class AuthRepository {
    private static final String ROLE_ELDERLY = "elderly";

    private final FirebaseAuth auth;
    private final DatabaseReference db;

    public AuthRepository() {
        auth = FirebaseAuth.getInstance();
        db = FirebaseDatabase.getInstance().getReference();
    }

    public boolean hasCurrentUser() {
        return auth.getCurrentUser() != null;
    }

    public void signOut() {
        auth.signOut();
    }

    public void requireCurrentElderly(RoleCallback callback) {
        if (auth.getCurrentUser() == null) {
            callback.onFailure("로그인이 필요합니다.");
            return;
        }

        loadCurrentUserRole(callback);
    }

    public void signInElderlyOnly(String email, String password, RoleCallback callback) {
        auth.signInWithEmailAndPassword(email, password)
                .addOnSuccessListener(result -> {
                    if (result.getUser() == null) {
                        callback.onFailure("사용자 정보를 확인하지 못했습니다.");
                        return;
                    }

                    loadCurrentUserRole(callback);
                })
                .addOnFailureListener(error ->
                        callback.onFailure(error.getMessage() == null ? "로그인에 실패했습니다." : error.getMessage())
                );
    }

    private void loadCurrentUserRole(RoleCallback callback) {
        String uid = auth.getCurrentUser() == null ? null : auth.getCurrentUser().getUid();

        if (uid == null) {
            callback.onFailure("로그인이 필요합니다.");
            return;
        }

        db.child("users")
                .child(uid)
                .child("role")
                .get()
                .addOnSuccessListener(snapshot -> {
                    String role = snapshot.getValue(String.class);

                    if (role == null) {
                        auth.signOut();
                        callback.onFailure("회원 역할 정보를 찾지 못했습니다.");
                        return;
                    }

                    if (!ROLE_ELDERLY.equals(role.trim().toLowerCase())) {
                        auth.signOut();
                        callback.onFailure("워치에서는 어르신 계정만 로그인할 수 있습니다.");
                        return;
                    }

                    callback.onSuccess(role);
                })
                .addOnFailureListener(error -> {
                    callback.onFailure(error.getMessage() == null ? "회원 정보를 불러오지 못했습니다." : error.getMessage());
                });
    }

    public interface RoleCallback {
        void onSuccess(String role);

        void onFailure(String message);
    }
}

File: app/src/main/java/com/example/next_contest/data/auth/AuthRepository.kt
package com.example.next_contest.data.auth

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ServerValue

data class AuthUserProfile(
    val uid: String,
    val email: String,
    val name: String,
    val phoneNumber: String,
    val normalizedPhone: String,
    val role: String
)

class AuthRepository(
    private val auth: FirebaseAuth = FirebaseAuth.getInstance(),
    private val db: DatabaseReference = FirebaseDatabase.getInstance().reference
) {

    fun getCurrentUserUid(): String? {
        return auth.currentUser?.uid
    }

    fun signOut() {
        auth.signOut()
    }

    fun signIn(
        email: String,
        password: String,
        onSuccess: (uid: String) -> Unit,
        onFailure: (message: String) -> Unit
    ) {
        auth.signInWithEmailAndPassword(email, password)
            .addOnSuccessListener { result ->
                val uid = result.user?.uid

                if (uid == null) {
                    onFailure("사용자 정보를 확인하지 못했습니다.")
                    return@addOnSuccessListener
                }

                onSuccess(uid)
            }
            .addOnFailureListener { error ->
                onFailure(error.message ?: "알 수 없는 오류")
            }
    }

    fun createUser(
        email: String,
        password: String,
        onSuccess: (uid: String) -> Unit,
        onFailure: (message: String) -> Unit
    ) {
        auth.createUserWithEmailAndPassword(email, password)
            .addOnSuccessListener { result ->
                val uid = result.user?.uid

                if (uid == null) {
                    onFailure("회원가입 정보를 확인하지 못했습니다.")
                    return@addOnSuccessListener
                }

                onSuccess(uid)
            }
            .addOnFailureListener { error ->
                onFailure(error.message ?: "알 수 없는 오류")
            }
    }

    fun loadUserRole(
        uid: String,
        onSuccess: (role: String) -> Unit,
        onFailure: (message: String) -> Unit
    ) {
        db.child("users")
            .child(uid)
            .child("role")
            .get()
            .addOnSuccessListener { roleSnap ->
                val role = roleSnap.getValue(String::class.java)

                if (role == null) {
                    onFailure("회원 정보를 찾지 못했습니다.")
                    return@addOnSuccessListener
                }

                onSuccess(role)
            }
            .addOnFailureListener { error ->
                onFailure(error.message ?: "알 수 없는 오류")
            }
    }

    fun isPhoneRegistered(
        normalizedPhone: String,
        onSuccess: (exists: Boolean) -> Unit,
        onFailure: (message: String) -> Unit
    ) {
        db.child("phoneIndex")
            .child(normalizedPhone)
            .get()
            .addOnSuccessListener { phoneSnap ->
                onSuccess(phoneSnap.exists())
            }
            .addOnFailureListener { error ->
                onFailure(error.message ?: "알 수 없는 오류")
            }
    }

    fun saveUserProfile(
        profile: AuthUserProfile,
        onSuccess: () -> Unit,
        onFailure: (message: String) -> Unit
    ) {
        val updates = hashMapOf<String, Any>(
            "users/${profile.uid}/uid" to profile.uid,
            "users/${profile.uid}/email" to profile.email,
            "users/${profile.uid}/name" to profile.name,
            "users/${profile.uid}/phoneNumber" to profile.phoneNumber,
            "users/${profile.uid}/normalizedPhone" to profile.normalizedPhone,
            "users/${profile.uid}/role" to profile.role,
            "users/${profile.uid}/createdAt" to ServerValue.TIMESTAMP,
            "phoneIndex/${profile.normalizedPhone}" to profile.uid
        )

        db.updateChildren(updates)
            .addOnSuccessListener {
                onSuccess()
            }
            .addOnFailureListener { error ->
                onFailure(error.message ?: "알 수 없는 오류")
            }
    }

    fun deleteCurrentUserThenSignOut() {
        val user = auth.currentUser

        if (user == null) {
            auth.signOut()
            return
        }

        user.delete()
            .addOnCompleteListener {
                auth.signOut()
            }
    }

}
